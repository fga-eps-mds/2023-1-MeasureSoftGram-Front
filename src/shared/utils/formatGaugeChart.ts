import { Historical } from '@customTypes/repository';

export interface FormatGaugeChartType {
  historical?: Historical[];
  title: string;
  isEmpty: boolean;
}

const formatGaugeChart = ({ historical, title, isEmpty }: FormatGaugeChartType) => {
  const maxSeriesItems:number = 6;
  const historicalLength: number = (historical?.length ?? 0) > maxSeriesItems ? maxSeriesItems : historical?.length ?? 0;
  const halfHistoricalLength:number = historicalLength / 2;
  let counter:number = halfHistoricalLength * -1;

  const seriesData = historical?.slice(0, maxSeriesItems).map((item) => {
    if (counter === 0) counter = 1;
    const offsetCenterX = `${counter * 80 }%`;
    counter += 1;
    return {
      value: item.latest.value.toFixed(2),
      name: item.name,
      title: {
        offsetCenter: [offsetCenterX, '80%']
      },
      detail: {
        offsetCenter: [offsetCenterX, '95%']
      }
    }
  }) ?? [];

  return {
    title: {
      text: title
    },
    series: {
      show: !isEmpty,
      min: 0,
      max: 1,
      data: seriesData,
      type: 'gauge',
      anchor: {
        show: true,
        showAbove: true,
        size: 18,
        itemStyle: {
          color: '#FAC858'
        }
      },
      pointer: {
        icon: 'path://M2.9,0.7L2.9,0.7c1.4,0,2.6,1.2,2.6,2.6v115c0,1.4-1.2,2.6-2.6,2.6l0,0c-1.4,0-2.6-1.2-2.6-2.6V3.3C0.3,1.9,1.4,0.7,2.9,0.7z',
        width: 8,
        length: '80%',
        offsetCenter: [0, '8%']
      },
      progress: {
        show: true,
        overlap: true,
        roundCap: true
      },
      axisLine: {
        roundCap: true
      },
      detail: {
        width: 20,
        height: 14,
        fontSize: 18,
        color: '#fff',
        backgroundColor: 'inherit',
        borderRadius: 3,
      }
    }

  }
};

export default formatGaugeChart;
