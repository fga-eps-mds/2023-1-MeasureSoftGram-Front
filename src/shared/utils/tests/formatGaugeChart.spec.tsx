import { Historical } from '@customTypes/repository';
import formatGaugeChart, { FormatGaugeChartType } from '@utils/formatGaugeChart';

describe('formatGaugeChart', () => {
  it('should return an object with the correct format', () => {
    const historic: Historical = {
      id: 1,
      key: 'reliability',
      name: 'Reliability',
      history: [],
      latest: {
        id: 12,
        value: 32,
        created_at: new Date(),
      }
    }
    const historical: Historical[] = [
      historic,
      historic,
      historic,
      historic
    ];

    const params: FormatGaugeChartType = {
      historical,
      title: 'title-test',
      isEmpty: false
    }

    const values = formatGaugeChart(params);

    expect(values).toMatchSnapshot();
  });

  it('should return an object with the correct format when historical is greater than 6', () => {
    const historic: Historical = {
      id: 1,
      key: 'reliability',
      name: 'Reliability',
      history: [],
      latest: {
        id: 12,
        value: 32,
        created_at: new Date(),
      }
    }
    const historical: Historical[] = [
      historic,
      historic,
      historic,
      historic,
      historic,
      historic,
      historic
    ];

    const params: FormatGaugeChartType = {
      historical,
      title: 'title-test',
      isEmpty: false
    }

    const values = formatGaugeChart(params);

    expect(values).toMatchSnapshot();
  });

  it('should return an object considering empty historic', () => {

    const params: FormatGaugeChartType = {
      historical: undefined,
      title: 'title-test',
      isEmpty: false
    }

    const values = formatGaugeChart(params);

    expect(values).toMatchSnapshot();
  });
});
