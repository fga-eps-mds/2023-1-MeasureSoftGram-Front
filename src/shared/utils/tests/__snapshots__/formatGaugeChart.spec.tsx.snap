// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formatGaugeChart should return an object considering empty historic 1`] = `
Object {
  "series": Object {
    "anchor": Object {
      "itemStyle": Object {
        "color": "#FAC858",
      },
      "show": true,
      "showAbove": true,
      "size": 18,
    },
    "axisLine": Object {
      "roundCap": true,
    },
    "data": Array [],
    "detail": Object {
      "backgroundColor": "inherit",
      "borderRadius": 3,
      "color": "#fff",
      "fontSize": 18,
      "height": 14,
      "width": 20,
    },
    "max": 1,
    "min": 0,
    "pointer": Object {
      "icon": "path://M2.9,0.7L2.9,0.7c1.4,0,2.6,1.2,2.6,2.6v115c0,1.4-1.2,2.6-2.6,2.6l0,0c-1.4,0-2.6-1.2-2.6-2.6V3.3C0.3,1.9,1.4,0.7,2.9,0.7z",
      "length": "80%",
      "offsetCenter": Array [
        0,
        "8%",
      ],
      "width": 8,
    },
    "progress": Object {
      "overlap": true,
      "roundCap": true,
      "show": true,
    },
    "show": true,
    "type": "gauge",
  },
  "title": Object {
    "text": "title-test",
  },
}
`;

exports[`formatGaugeChart should return an object with the correct format 1`] = `
Object {
  "series": Object {
    "anchor": Object {
      "itemStyle": Object {
        "color": "#FAC858",
      },
      "show": true,
      "showAbove": true,
      "size": 18,
    },
    "axisLine": Object {
      "roundCap": true,
    },
    "data": Array [
      Object {
        "detail": Object {
          "offsetCenter": Array [
            "-160%",
            "95%",
          ],
        },
        "name": "Reliability",
        "title": Object {
          "offsetCenter": Array [
            "-160%",
            "80%",
          ],
        },
        "value": "32.00",
      },
      Object {
        "detail": Object {
          "offsetCenter": Array [
            "-80%",
            "95%",
          ],
        },
        "name": "Reliability",
        "title": Object {
          "offsetCenter": Array [
            "-80%",
            "80%",
          ],
        },
        "value": "32.00",
      },
      Object {
        "detail": Object {
          "offsetCenter": Array [
            "80%",
            "95%",
          ],
        },
        "name": "Reliability",
        "title": Object {
          "offsetCenter": Array [
            "80%",
            "80%",
          ],
        },
        "value": "32.00",
      },
      Object {
        "detail": Object {
          "offsetCenter": Array [
            "160%",
            "95%",
          ],
        },
        "name": "Reliability",
        "title": Object {
          "offsetCenter": Array [
            "160%",
            "80%",
          ],
        },
        "value": "32.00",
      },
    ],
    "detail": Object {
      "backgroundColor": "inherit",
      "borderRadius": 3,
      "color": "#fff",
      "fontSize": 18,
      "height": 14,
      "width": 20,
    },
    "max": 1,
    "min": 0,
    "pointer": Object {
      "icon": "path://M2.9,0.7L2.9,0.7c1.4,0,2.6,1.2,2.6,2.6v115c0,1.4-1.2,2.6-2.6,2.6l0,0c-1.4,0-2.6-1.2-2.6-2.6V3.3C0.3,1.9,1.4,0.7,2.9,0.7z",
      "length": "80%",
      "offsetCenter": Array [
        0,
        "8%",
      ],
      "width": 8,
    },
    "progress": Object {
      "overlap": true,
      "roundCap": true,
      "show": true,
    },
    "show": true,
    "type": "gauge",
  },
  "title": Object {
    "text": "title-test",
  },
}
`;

exports[`formatGaugeChart should return an object with the correct format when historical is greater than 6 1`] = `
Object {
  "series": Object {
    "anchor": Object {
      "itemStyle": Object {
        "color": "#FAC858",
      },
      "show": true,
      "showAbove": true,
      "size": 18,
    },
    "axisLine": Object {
      "roundCap": true,
    },
    "data": Array [
      Object {
        "detail": Object {
          "offsetCenter": Array [
            "-240%",
            "95%",
          ],
        },
        "name": "Reliability",
        "title": Object {
          "offsetCenter": Array [
            "-240%",
            "80%",
          ],
        },
        "value": "32.00",
      },
      Object {
        "detail": Object {
          "offsetCenter": Array [
            "-160%",
            "95%",
          ],
        },
        "name": "Reliability",
        "title": Object {
          "offsetCenter": Array [
            "-160%",
            "80%",
          ],
        },
        "value": "32.00",
      },
      Object {
        "detail": Object {
          "offsetCenter": Array [
            "-80%",
            "95%",
          ],
        },
        "name": "Reliability",
        "title": Object {
          "offsetCenter": Array [
            "-80%",
            "80%",
          ],
        },
        "value": "32.00",
      },
      Object {
        "detail": Object {
          "offsetCenter": Array [
            "80%",
            "95%",
          ],
        },
        "name": "Reliability",
        "title": Object {
          "offsetCenter": Array [
            "80%",
            "80%",
          ],
        },
        "value": "32.00",
      },
      Object {
        "detail": Object {
          "offsetCenter": Array [
            "160%",
            "95%",
          ],
        },
        "name": "Reliability",
        "title": Object {
          "offsetCenter": Array [
            "160%",
            "80%",
          ],
        },
        "value": "32.00",
      },
      Object {
        "detail": Object {
          "offsetCenter": Array [
            "240%",
            "95%",
          ],
        },
        "name": "Reliability",
        "title": Object {
          "offsetCenter": Array [
            "240%",
            "80%",
          ],
        },
        "value": "32.00",
      },
    ],
    "detail": Object {
      "backgroundColor": "inherit",
      "borderRadius": 3,
      "color": "#fff",
      "fontSize": 18,
      "height": 14,
      "width": 20,
    },
    "max": 1,
    "min": 0,
    "pointer": Object {
      "icon": "path://M2.9,0.7L2.9,0.7c1.4,0,2.6,1.2,2.6,2.6v115c0,1.4-1.2,2.6-2.6,2.6l0,0c-1.4,0-2.6-1.2-2.6-2.6V3.3C0.3,1.9,1.4,0.7,2.9,0.7z",
      "length": "80%",
      "offsetCenter": Array [
        0,
        "8%",
      ],
      "width": 8,
    },
    "progress": Object {
      "overlap": true,
      "roundCap": true,
      "show": true,
    },
    "show": true,
    "type": "gauge",
  },
  "title": Object {
    "text": "title-test",
  },
}
`;
